R/estimateDominance.R

```diff
@@ -0,0 +1,366 @@
+#'
+#' @param name A name for the column of the colData where the calculated
+#' Dominance indices should be stored in.
+#'
+#' @param BPPARAM A
+#'   \code{\link[BiocParallel:BiocParallelParam-class]{BiocParallelParam}}
+#'   object specifying whether calculation of estimates should be parallelized.
+#'
+#' @details
+#' \code{estimateDominance} calculates the community dominance indices.
+#'
+#' @return Returns ME object, that includes dominance indices in coldata.
+#'
+#' @seealso
+#' \itemize{
+#'   \item{\code{\link[=estimateDiversity]{estimateDiversity}}}
+#' }
+#'
+#' @name estimateDominance
+#' @export
+#'
+#' @author Leo Lahti and Tuomas Borman. Contact: \url{microbiome.github.io}
+#'
+#' @examples
+#' data(esophagus, package = "MicrobiomeExperiment")
+#' esophagus <- as(esophagus, "MicrobiomeExperiment")
+#'
+#' #Calculates simpson dominance index
+#' esophagus <- estimateDominance(esophagus, index="simpson")
+#' #Shows all indices
+#' colData(esophagus)
+#'
+#' #Tries to calculate index that is not accepted. Gets an error.
+#' esophagus <- estimateDominance(esophagus, index="does_not_exist")
+#' #Tries to calculate index that is not accepted
+#' # and index that is accepted. Gets an error.
+#' esophagus <- estimateDominance(esophagus, index=c("dmn", "does_not_exist"))
+#' #Shows all indices
+#' colData(esophagus)
+#'
+#' #Calculates DBP and Core Abundance indices
+#' esophagus <- estimateDominance(esophagus, index=c("dbp", "core_abundance"))
+#' #Shows all indices
+#' colData(esophagus)
+#' #Shows DBP index
+#' colData(esophagus)$dbp
+#'
+#' #Deletes DBP index
+#' colData(esophagus)$dbp <- NULL
+#' #Shows all indices, DBP is deleted
+#' colData(esophagus)
+#' #Deletes all indices
+#' colData(esophagus) <- NULL
+#'
+#' #Names of columns can be chosen, but the length of arguments must match.
+#' esophagus <- estimateDominance(esophagus,
+#'     index=c("dbp", "core_abundance"),
+#'     name = c("index1", "index2"))
+#' #Shows all indices
+#' colData(esophagus)
+#' #If they do not match, gets an error.
+#' esophagus <- estimateDominance(esophagus,
+#'     index="simpson",
+#'     name = c("index3", "index4"))
+#' #Shows all indices
+#' colData(esophagus)
+#' #Deletes all indices
+#' colData(esophagus) <- NULL
+#'
+#' #Calculates all indices
+#' esophagus <- estimateDominance(esophagus)
+#' #Or you can calculate all indices by specify "all"
+#' esophagus <- estimateDominance(esophagus, index="all")
+#' #Shows all indices
+#' colData(esophagus)
+#'
+#' @export
+#'
+setGeneric("estimateDominance",signature = c("x"),
+           function(x, abund_values = "counts", index="all", rank=1, as_relative=TRUE, aggregate=TRUE,
+                    name = index, ...)
+               standardGeneric("estimateDominance"))
+
+
+#' @rdname estimateDominance
+#' @export
+setMethod("estimateDominance", signature = c(x = "MicrobiomeExperiment"),
+          function(x, abund_values = "counts", index="all", rank=1, as_relative=TRUE, aggregate=TRUE,
```

Comment:
I agree, that it would be beneficial to use that simple solution. Although with this, I think it is not possible to calculate multiple indices at the same time. With the .check_indices, it was. 

Is it ok, that it is possible to calculate indices only one by one?

(Well, e.g. EstimateDiversity's indices are calculated also one by one, because its function are like estimateRichness. estimateShannon...)

Reply:
Hmm.. but `estimateDiversity(x)` does calculate all indices by default (see 3978bbf). Isn't this what you are after as well?

Reply:
just do a `match.arg(index, several.ok = TRUE)`. Then you get a vector back, which can only contain values as defined in the function definition. If you then do a lapply on those you should be able to get all the values you need. The lapply can also be switched to `BiocParallel::bplapply` so that the values for the different indices are calculated in parallel

Reply:
You can have a look at `estimateDiversity` including the latest changes, if you want to implement parallelization. However, parallelization makes only sense, if there are more then 1000 samples and mostly on non-windows systems.

Reply:
i.m.o would make sense to implement it since large collections are increasingly common now

Reply:
Ok, thanks for clarifying. I replaced the check_indices function with this more simple solution

Reply:
Implemented parallelization.